segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/fs/FSStorage.java

```diff
@@ -0,0 +1,378 @@
+        return CompletableFuture.supplyAsync(() -> {
+            Path path = Paths.get(config.getNfsRoot(), streamSegmentName);
+            if (!Files.exists(path)) {
+                throw new CompletionException(new StreamSegmentNotExistsException(streamSegmentName));
+            } else if (Files.isWritable(path)) {
+                return FSSegmentHandle.getWriteHandle(streamSegmentName);
+            } else {
+                try {
+                    return openRead(streamSegmentName).get();
+                } catch (InterruptedException | ExecutionException e) {
+                    throw new CompletionException(e);
+                }
+            }
+        }, executor);
+    }
+
+    @Override
+    public CompletableFuture<SegmentProperties> create(String streamSegmentName, Duration timeout) {
+        return CompletableFuture.supplyAsync( () -> syncCreate(streamSegmentName, timeout), executor);
+    }
+
+    @Override
+    public CompletableFuture<Void> write(SegmentHandle handle, long offset, InputStream data, int length, Duration
+            timeout) {
+        return CompletableFuture.supplyAsync(() -> syncWrite(handle, offset, data, length, timeout), executor);
+    }
+
+    @Override
+    public CompletableFuture<Void> seal(SegmentHandle handle, Duration timeout) {
+        return CompletableFuture.supplyAsync(() -> syncSeal(handle, timeout), executor);
+    }
+
+    @Override
+    public CompletableFuture<Void> concat(SegmentHandle targetHandle, long offset, String sourceSegment, Duration
+            timeout) {
+            return CompletableFuture.supplyAsync(() -> syncConcat(targetHandle, offset, sourceSegment, timeout),
+                    executor);
+    }
+
+    @Override
+    public CompletableFuture<Void> delete(SegmentHandle handle, Duration timeout) {
+            return CompletableFuture.supplyAsync(() -> syncDelete(handle, timeout), executor);
+    }
+
+    //endregion
+
+    //region AutoClosable
+
+    @Override
+    public void close() {
+
+    }
+
+    //endregion
+
+    //region private sync implementation
+
+    private SegmentHandle syncOpenRead(String streamSegmentName) {
+        Path path = Paths.get(config.getNfsRoot(), streamSegmentName);
+
+        if (!Files.exists(path)) {
+            throw new CompletionException(new StreamSegmentNotExistsException(streamSegmentName));
+        }
+
+        FSSegmentHandle retHandle = FSSegmentHandle.getReadHandle(streamSegmentName);
+        return retHandle;
+    }
+
+
+    private int syncRead(SegmentHandle handle, long offset, byte[] buffer, int bufferOffset, int length, Duration
+            timeout) {
+
+        Path path = Paths.get(config.getNfsRoot(), handle.getSegmentName());
+
+        if (!Files.exists(path)) {
+            throw new CompletionException(new StreamSegmentNotExistsException(handle.getSegmentName(), null));
+        }
+
+        try {
+            if (Files.size(path) < offset) {
```

Comment:
Wrong Exception Type. `ArrayIndexOutOfBounds` is when you have an array.  You need to throw `IllegalArgumentException`. Please refer to `HDFSStorage`/`ReadOperation` for an example of argument validation.